clients/src/main/java/org/apache/kafka/common/network/SaslChannelBuilder.java

```diff
@@ -222,4 +268,30 @@ private static String defaultKerberosRealm() throws ClassNotFoundException, NoSu
         getDefaultRealmMethod = classRef.getDeclaredMethod("getDefaultRealm", new Class[0]);
         return (String) getDefaultRealmMethod.invoke(kerbConf, new Object[0]);
     }
+
+    private void createClientCallbackHandler(Map<String, ?> configs, String mechanism) {
+        Class<? extends AuthenticateCallbackHandler> clazz = (Class<? extends AuthenticateCallbackHandler>) configs.get(SaslConfigs.SASL_CLIENT_CALLBACK_HANDLER_CLASS);
+        if (clazz == null)
+            clazz = mechanism.equals(SaslConfigs.GSSAPI_MECHANISM) ? KerberosClientCallbackHandler.class : SaslClientCallbackHandler.class;
+        AuthenticateCallbackHandler callbackHandler = Utils.newInstance(clazz);
+        saslCallbackHandlers.put(mechanism, callbackHandler);
+    }
+
+    private void createServerCallbackHandlers(Map<String, ?> configs, List<String> enabledMechanisms) throws ClassNotFoundException {
+        String callbackClasses = (String) configs.get(BrokerSecurityConfigs.SASL_SERVER_CALLBACK_HANDLER_CLASS_MAP);
+        Map<String, String> callbackClassMap = Utils.parseCsvMap(callbackClasses);
+        for (String mechanism : enabledMechanisms) {
+            AuthenticateCallbackHandler callbackHandler;
+            String className = callbackClassMap.get(mechanism);
+            if (className != null)
+                callbackHandler = Utils.newInstance(className, AuthenticateCallbackHandler.class);
+            else if (mechanism.equals(PlainSaslServer.PLAIN_MECHANISM))
```

Comment:
Would it ever be acceptable to use the config map to override the callback handlers for the built-in mechanism implementations (GSSAPI, PLAIN, and the two SCRAM-related ones)?  If so then the code is fine, but if not then the built-in mechanism names should be explicitly checked for first.  Same goes for the client side checks above.

Reply:
@rondagostino Thanks for the review. Yes, it is acceptable to override callback handlers for built-in mechanisms. One of the motivations for the KIP is to enable customization of built-in mechanisms to integrate with existing authentication servers (e.g. use an existing server or database to verify passwords for PLAIN).